pkg/reconciler/route/traffic/rollout.go

```diff
@@ -70,3 +73,100 @@ type RevisionRollout struct {
 	// target percentage.
 	Percent int `json:"percent"`
 }
+
+// Step processes previous Rollout object and compares to the current
+// rollout state. If there is different, Step will start or stop the rollout
+// and updates `RevisionRollout` objects of `cur` accoringly.
+// At the end of the call `cur` contains the desired traffic shape.
+// Step returns `true` if any changes have been made.
+func (cur *Rollout) Step(prev *Rollout) bool {
+	if prev == nil {
+		return false
+	}
+
+	// The algorithm below is simplest, but probably not the most performant.
+	// Optimize in the later passes.
+	// Map by tag.
+	currConfigs, prevConfigs := map[string][]*ConfigurationRollout{}, map[string][]*ConfigurationRollout{}
+	for i, cfg := range cur.Configurations {
+		currConfigs[cfg.Tag] = append(currConfigs[cfg.Tag], &cur.Configurations[i])
+	}
+	for i, cfg := range prev.Configurations {
+		prevConfigs[cfg.Tag] = append(prevConfigs[cfg.Tag], &prev.Configurations[i])
+	}
+	ret := false
+	for t, ccfgs := range currConfigs {
+		pcfgs, ok := prevConfigs[t]
+		// A new tag was added, so we have no previous state to roll from,
+		// thus just move over, we'll rollout to 100% from the get go (and it is
+		// always 100%, since default tag is _always_ there).
+		if !ok {
+			continue
+		}
+		// This is basically an intersect algorithm,
+		// It relies on the fact that inputs are sorted.
+		for i, j := 0, 0; i < len(ccfgs) && j < len(pcfgs); {
+			switch {
+			case ccfgs[i].ConfigurationName == pcfgs[j].ConfigurationName:
+				// Config might have 0% traffic assigned, if it is a tag only route (i.e.
+				// receives no traffic via default tag).
+				if ccfgs[i].Percent != 0 {
+					ret = ret || stepConfig(ccfgs[i], pcfgs[j])
+				}
+				i++
+				j++
+			case ccfgs[i].ConfigurationName < pcfgs[j].ConfigurationName:
+				// A new config, has been added. No action for rollout though.
+				i++
+			default: // cur > prev.
+				// A config has been removed. Again, no action for rollout, since
+				// this will no longer be rolling out even it were.
+				j++
```

Comment:
If I move a tag from config A to config B, I think this will skip right over it, no?

Reply:
sunk up offline, it's the fact that this remains in the current object, which is subtle.

I think functional will make this clearer.